logger/logger.go

```diff
@@ -0,0 +1,166 @@
+package logger
+
+import (
+	"fmt"
+	"github.com/lestrrat-go/file-rotatelogs"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cast"
+	"github.com/spf13/viper"
+	"io"
+	"os"
+	"time"
+)
+
+// Logger configuration keys
+const (
+	LogLevelKey     = "level"
+	LogFormatterKey = "formatter"
+	LogOutputKey    = "output"
+
+	LogFormatterTypeKey     = "type"
+	LogFormatterTimezoneKey = "timezone"
+
+	LogOutputTypeKey                  = "type"
+	LogOutputFileFilePatternKey       = "file_pattern"
+	LogOutputFileCurrentLinkKey       = "current_link"
+	LogOutputFileClockTimezoneKey     = "clock_timezone"
+	LogOutputFileRotationTimeInSecKey = "rotation_time_in_sec"
+	LogOutputFileMaxAgeInSecKey       = "max_age_in_sec"
+	LogOutputFileRotationCountKey     = "rotation_count"
+)
+
+// InitLogger initializes logger using configuration provided by viper
+// instance.
+//
+// Function designed to configure few different loggers with different
+// formatter and output settings. To achieve this viper configuration
+// contains separate sections for each logger, each output and
+// each formatter. As viper doesn't support defaults for sub-configurations
+// there is a workaround for this. Two different viper configurations should be
+// provided: one for main configuration and one for configuration defaults.
+func InitLogger(config *viper.Viper, defaults *viper.Viper) error {
+	return initLogger(&configWithDefaults{config, defaults})
+}
+
+type configWithDefaults struct {
+	config   *viper.Viper
+	defaults *viper.Viper
+}
+
+func (config *configWithDefaults) get(key string) interface{} {
+	if config.config != nil && config.config.InConfig(key) {
+		return config.config.Get(key)
+	}
+	return config.defaults.Get(key)
+}
+
+func (config *configWithDefaults) getString(key string) string {
+	return cast.ToString(config.get(key))
+}
+
+func (config *configWithDefaults) getInt(key string) int {
+	return cast.ToInt(config.get(key))
+}
+
+func (config *configWithDefaults) getDuration(key string) time.Duration {
+	return cast.ToDuration(config.get(key))
+}
+
+func (config *configWithDefaults) sub(key string) *configWithDefaults {
+	return &configWithDefaults{config.config.Sub(key), config.defaults.Sub(key)}
+}
+
+func initLogger(config *configWithDefaults) error {
+	var err error
+
+	var level log.Level
+	var levelString = config.getString(LogLevelKey)
+	level, err = log.ParseLevel(levelString)
+	if err != nil {
+		return fmt.Errorf("Unable parse log level string: %v, err: %v", levelString, err)
+	}
+	log.SetLevel(level)
+
+	var formatter log.Formatter
+	formatter, err = newFormatterByConfig(config.sub(LogFormatterKey))
+	if err != nil {
+		return fmt.Errorf("Unable initialize log formatter, error: %v", err)
+	}
+	log.SetFormatter(formatter)
+
+	var output io.Writer
+	output, err = newOutputByConfig(config.sub(LogOutputKey))
+	if err != nil {
+		return fmt.Errorf("Unable initialize log output, error: %v", err)
+	}
+	log.SetOutput(output)
```

Comment:
Should the output writer be closed on the app side in some way or it is handled by the logrus?